### PR TITLE
Authentication via Northstar.

### LINKED ISSUE #3: admin login
via NS

### PR DESCRIPTION
**Pull request by weerd**
#### What's this PR do?
This PR enables authentication for Gladiator via Northstar API. Only users that have a role of `admin` or `staff` can access authenticated routes. This also creates the login view to allow authorized users to login. 

#### Where should the reviewer start?
Probably start in the **AuthServiceProvider.php** then head to the **GladiatorUserProvider.php** etc. Checkout the **routes.php** too. There's a lot of stuff... I apologize ahead of time :grimacing: 

#### How should this be manually tested?
You definitely need to update your **.env** to now include:
```
NORTHSTAR_URI=https://northstar.dosomething.org
NORTHSTAR_API_KEY=GrabTheKeyFromAurora
```
You can then use the artisan console command to add yourself as a user to the database and attempt to login:

```
$ php artisan add:user username@dosomething.org --role=admin
```

#### Any background context you want to provide?
Make sure to run the following commands if testing this branch or once this branch is merged:
```
$ composer dump-autoload
$ php artisan clear-compiled
```

I still need to populate the message bag to give proper validation errors if a user can't login for whatever reason. There's a couple smaller updates coming in a later PR, just didn't want this one to get too massive.

#### What are the relevant tickets?
Fixes #3 

#### Screenshots (if appropriate)
This PR also adds a convenient artisan command called `add:user` which allows you to create a new user with a role of `admin` or `staff` ("staff" is the default if none is specified). These users are first verified as DoSomething users on Northstar before they are added via a specified email:

```
$ php artisan add:user dlorenzo@dosomething.org --role=admin
```

screen shot 2016-02-24 at 2 08 28 am

It checks to make sure they are in Northstar and user isn't already in the Gladiator database.

This will help with local development when you need to authenticate into the app, and also when we deploy to production, we could have an artisan command run on initial deploy to officially add new admin and staff users that can automatically begin using the site from the ["gecko"](http://www.urbandictionary.com/define.php?term=from+the+gecko)

---
@angaither @sbsmith86 @deadlybutter 

cc: @DFurnes

### CODE REVIEW

**Comment by weerd** [comment]
I'm rebasing and then fixing StyleCI stuff!

File: app/Services/RestApiClient.php [hunk 1]
```diff
@@ -0,0 +1,138 @@
+<?php
+
+namespace Gladiator\Services;
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\RequestException;
+
+class RestApiClient
```
Comment (sbsmith86) [inline app/Services/RestApiClient.php]: Is this in place of using something like guzzle?

Comment (sbsmith86) [inline app/Services/RestApiClient.php]: whoa brain fart. Sorry. ignore this.

**Comment by DFurnes** [comment]
> You definitely need to update your __.env__ to now include:
> ```
NORTHSTAR_URI=https://northstar.dosomething.org
NORTHSTAR_API_KEY=GrabTheKeyFromAurora
```

For local dev, I think it'd be a good idea to use [our QA instance](http://qa-aurora.dosomething.org) or a local Northstar environment so we don't accidentally muck anything up in production.

**Comment by weerd** [comment]
@DFurnes it's not writing anything to Northstar just checking stuff, so would that still be an issue?

File: app/Services/RestApiClient.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * RestApiClient constructor.
+     *
+     * @param string  $base_uri
+     * @param array  $headers
+     */
+    public function __construct($base_uri, $headers = [])
+    {
+        $headers += [
```
Comment (DFurnes) [inline app/Services/RestApiClient.php]: :eyes:

**Comment by deadlybutter** [comment]
@DFurnes Id rather go with using the QA instance just cause its one less thing you need to setup when starting dev

File: app/Http/routes.php [hunk 3]
```diff
@@ -10,14 +11,22 @@
...
+    Route::get('auth/login', 'Auth\AuthController@showLoginForm');
+    Route::post('auth/login', 'Auth\AuthController@login');
+    Route::get('auth/logout', 'Auth\AuthController@logout');
 
     // Waiting rooms routes.
     Route::model('waitingrooms', 'Gladiator\Models\WaitingRoom');
     Route::resource('waitingrooms', 'WaitingRoomsController');
+
+    // Competitions
+    Route::resource('competitions', 'CompetitionsController');
```
Comment (angaither) [inline app/Http/routes.php]: ocd says put this above waiting rooms for reasons of :abc:

**Comment by weerd** [comment]
@deadlybutter you're not actually setting up Northstar locally. It's requesting from the official Northstar. If anything it's good for testing it out ;)

**Comment by DFurnes** [comment]
@weerd Hmm, yeah I guess if you remove the `admin` and `user` scopes from the key we could do that. I'd still vote for QA just so we can have the credentials in the `.env.example` like @deadlybutter suggested above.

File: app/Services/RestApiClient.php [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        return json_decode($response->getBody(), $asArray)->data;
+    }
+
+    /**
+     * Make JSON object string from the supplied data.
+     *
+     * @param  object  $data
+     * @return string
+     */
+    public function makeJson($data)
```
Comment (angaither) [inline app/Services/RestApiClient.php]: lol

File: resources/views/layouts/master.blade.php [hunk 5]
```diff
@@ -34,6 +39,7 @@
             </div>
         </div>
 
+         <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
```
Comment (angaither) [inline resources/views/layouts/master.blade.php]: extra space :space_invader:

**Comment by weerd** [comment]
@DFurnes and I had an IRL chat and makes sense. I removed the `admin` scope in the Gladiator key and gonna switch the .env-example to use the QA instance info :)

File: app/Http/routes.php (hunk 3 above)
Comment (weerd) [inline app/Http/routes.php]: my ocd is surprised it didn't catch that earlier ;)

**Comment by weerd** [comment]
I'm fixing up some views stuff that is pretty broken. The rebase/merge happened, but didn't exactly resolve everything :grimacing:

File: app/Services/RestApiClient.php (hunk 2 above)
Comment (weerd) [inline app/Services/RestApiClient.php]: hehe :smile: adds to the array and if it's an empty array then doesn't add any items

**Comment by weerd** [comment]
@sbsmith86 @deadlybutter I had to make a bunch of updates to the templates because I had a slew of conflicts. Also found a bug or two. Just refer to the commit:
https://github.com/DoSomething/gladiator/commit/dfecbf5d78385df403223eb2e7d8ebf003bcf300

**Comment by sbsmith86** [comment]
@weerd sounds good.

**Comment by angaither** [comment]
`¯\_(ツ)_/¯` :+1:

**Pull request merged**